tower-defense/src/game_state/tower/mod.rs

```diff
@@ -70,11 +101,37 @@ impl Deref for Tower {
 pub struct TowerTemplate {
     pub kind: TowerKind,
     pub shoot_interval: Duration,
-    pub attack_range_radius: f32,
+    pub default_attack_range_radius: f32,
     pub projectile_kind: ProjectileKind,
     pub projectile_speed: Velocity,
-    pub damage: usize,
+    pub default_damage: usize,
 }
 
 #[derive(Clone, Copy)]
 pub enum TowerKind {}
+
+pub fn tower_cooldown_tick(game_state: &mut GameState, dt: Duration) {
+    game_state.towers.iter_mut().for_each(|tower| {
+        if tower.cooldown == Duration::from_secs(0) {
+            return;
+        }
+
+        let mut cooldown_sub = dt;
+        tower.status_effects.iter().for_each(|status_effect| {
+            if let TowerStatusEffectKind::AttackSpeedAdd { add } = status_effect.kind {
+                cooldown_sub += Duration::from_secs_f32(add);
+            }
+        });
```

Comment:
This is problematic. `add` is not second.

Reply:
How about using the following methods?

- On shoot, store shoot time and calculate next shoot time
- The next shoot time can be calculated as shoot time + (1 sec/attack speed)
   - At this point, the attack speed means how many times per second tower can attack
- If the attack speed buff is applied after launch, just recalculate the next shoot time

How do you think?

Reply:
I tried new way to calculate it. please check it again.